fluence-faas-service/src/service.rs

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::Result;
+use super::ServiceError;
+use super::IValue;
+
+use fluence_faas::FluenceFaaS;
+use fluence_faas::ModulesConfig;
+
+use std::convert::TryInto;
+
+const SERVICE_ID_ENV_NAME: &str = "service_id";
+
+// TODO: remove and use mutex instead
+unsafe impl Send for FluenceFaaSService {}
+
+pub struct FluenceFaaSService {
+    faas: FluenceFaaS,
+}
+
+impl FluenceFaaSService {
+    /// Creates Service with given modules and service id.
+    pub fn new<I, C, S>(modules: I, config: C, service_id: S) -> Result<Self>
+    where
+        I: IntoIterator<Item = String>,
+        C: TryInto<ModulesConfig>,
+        S: AsRef<str>,
+        ServiceError: From<C::Error>,
+    {
+        let config: ModulesConfig = config.try_into()?;
+        let service_id = service_id.as_ref();
+        let config = Self::prepare_before_creation(config, service_id)?;
+
+        let modules = modules.into_iter().collect();
+        let faas = FluenceFaaS::with_module_names(&modules, config)?;
+
+        Ok(Self { faas })
+    }
+
+    /// Call a specified function of loaded module by its name.
+    // TODO: replace serde_json::Value with Vec<u8>?
+    pub fn call_module<MN: AsRef<str>, FN: AsRef<str>>(
+        &mut self,
+        module_name: MN,
+        func_name: FN,
+        arguments: serde_json::Value,
+    ) -> Result<Vec<IValue>> {
+        let arguments = Self::prepare_arguments(arguments)?;
+
+        self.faas
+            .call_module(module_name, func_name, &arguments)
+            .map_err(Into::into)
+    }
+
+    /// Return all export functions (name and signatures) of loaded modules.
+    pub fn get_interface(&self) -> fluence_faas::FaaSInterface<'_> {
+        self.faas.get_interface()
+    }
+
+    fn prepare_before_creation(
```

Comment:
maybe call it `set_env_and_dirs`? It's kinda silly, but I don't like current name because it says almost nothing